src/Microsoft.OData.Client/ALinq/QueryableResourceExpression.cs

```diff
@@ -371,6 +371,56 @@ internal void AddFilter(IEnumerable<Expression> predicateConjuncts)
             this.keyPredicateConjuncts.Clear();
         }
 
+        internal void AddApply(Expression aggregateExpr, OData.UriParser.Aggregation.AggregationMethod aggregationMethod)
+        {
+            if (this.OrderBy != null)
+            {
+                throw new NotSupportedException(Strings.ALinq_QueryOptionOutOfOrder("apply", "orderby"));
+            }
+            else if (this.Skip != null)
+            {
+                // $skip and $top may be used together with rollup. 
+                // However, support for rollup is currently not implemented in OData WebApi
+                // If $skip and/or $top appears before $apply, its currently ignored. 
+                // Makes sense to throw an exception to avoid giving a false impression.
+                throw new NotSupportedException(Strings.ALinq_QueryOptionOutOfOrder("apply", "skip"));
+            }
+            else if (this.Take != null)
+            {
+                throw new NotSupportedException(Strings.ALinq_QueryOptionOutOfOrder("apply", "top"));
+            }
+
+            if (this.Apply == null)
+            {
+                AddSequenceQueryOption(new ApplyQueryOptionExpression(this.Type));
+            }
+
+            if (this.Filter != null && this.Filter.PredicateConjuncts.Count > 0)
+            {
+                // The $apply query option is evaluated first, then other query options ($filter, $orderby, $select) are evaluated, 
+                // if applicable, on the result of $apply in their normal order.
+                // http://docs.oasis-open.org/odata/odata-data-aggregation-ext/v4.0/cs02/odata-data-aggregation-ext-v4.0-cs02.html#_Toc435016590
+                
+                // If a Where appears before an aggregation method (e.g. Average, Sum, etc) or GroupBy, 
+                // the conjuncts of the filter expression will be used to restrict the set of data to be aggregated. 
+                // They will not appear on the $filter query option. Instead, we use them to construct a filter transformation.
+                // E.g. /Sales?$apply=filter(Amount gt 1)/aggregate(Amount with average as AverageAmount)
+                
+                // If a Where appears after an aggregation method or GroupBy, the conjuncts should appear
+                // on a $filter query option after the $apply.
+                // E.g. /Sales?$apply=groupby((Product/Color),aggregate(Amount with average as AverageAmount))&$filter=Product/Color eq 'Brown'

```

Comment:
So there's support for having Where clauses both before and after an aggregation? In this case the generated query would contain both a filter transformation in the $apply clause as well as a $filter clause? Like `/Sales?$apply=filter(Amount gt 1)/groupby((Product/Color),aggregae(Amount with average as Avg))$filter=Product/Color eq 'Brown'`?
Is this a correct implication of what you've explained here?

Reply:
@habbes You understood it correctly. A `$filter` (_as well as `$order` and `$select`_) query option can appear after the `$apply`. When it does, it'll be applied to the result of the aggregation.

As a matter of fact, a `filter` transformation can also appear after `groupby` and `aggregate` expression. When it does, it'll be applied to the result of the aggregation, such that the following two expressions give you the same result:

1. `/Sales?$apply=filter(Amount gt 1)/groupby((Product/Color),aggregate(Amount with average as AverageAmount))&$filter=Product/Color eq 'Brown'`
We're dealing with 2 query params here - `$apply` and `$filter` - separated by a `&`. The `$apply` here has 2 transformations - `filter` and `groupby` - separated by a forward slash (/). The `$filter` will be applied to the aggregated result.

2. `/Sales?$apply=filter(Amount gt 1)/groupby((Product/Color),aggregate(Amount with average as AverageAmount))/filter(Product/Color eq 'Brown')`
We're dealing with a single query param here - `$apply`. It has 3 transformations separated by a forward slash (/). The `filter` transformation before `groupby` restricts the set of data to be aggregated while the one that appears after is applied to the aggregated result

That said, I deliberately decided to leave out  support for `filter` transformation **appearing after** the `groupby` and `aggregate` transformation as well as system query options **appearing after** `$apply`. The effort required to make them play is not trivial. That work will tackled in a separate work item/pull request